cpp/Platform.Delegates/Platform.Delegates.Delegate.h

```diff
@@ -16,54 +14,134 @@ namespace Platform::Delegates
+            if (this != &other)
+            {
+                this->simpleFunction = other.simpleFunction;
+                this->memberMethod = other.memberMethod;
+                this->complexFunction = other.complexFunction;
+            }
+            return *this;
+        }
+
+        Delegate &operator=(Delegate&& other) noexcept {
+            if (this != &other)
+            {
+                this->simpleFunction = std::move(other.simpleFunction);
+                this->memberMethod = std::move(other.memberMethod);
+                this->complexFunction = std::move(other.complexFunction);
+            }
+            return *this;
+        }
+
+        virtual ReturnType operator()(Args... args)
+        {
+            if (simpleFunction)
+            {
+                return simpleFunction(args...);
+            }
+            if (memberMethod)
+            {
+                return (*memberMethod)(args...);
+            }
+            if (complexFunction)
+            {
+                return (*complexFunction)(args...);
+            }
+
+            throw std::bad_function_call{};
+        }
+
+        virtual bool operator==(const Delegate &other) const
+        {
+            if (simpleFunction && other.simpleFunction)
+            {
+                return simpleFunction == other.simpleFunction;
+            }
+            if (memberMethod && other.memberMethod)
+            {
+                return *memberMethod == *other.memberMethod;
+            }
+            if (complexFunction && other.complexFunction)
+            {
+                return AreFunctionsEqual(*complexFunction, *other.complexFunction);
+            }
+            return false;
+        }
+
+        virtual bool operator!=(const Delegate &other) const
+        {
+            return !(*this == other);
+        }
+    private:
         class MemberMethodBase
         {
         public:
             virtual ReturnType operator()(Args... args) = 0;
             virtual bool operator== (const MemberMethodBase &other) const = 0;
             virtual bool operator!= (const MemberMethodBase &other) const = 0;
+            virtual ~MemberMethodBase() = default;
         };
 
         template <typename Class>
         class MemberMethod : public MemberMethodBase
         {
-            std::shared_ptr<Class> object;
-            ReturnType(Class:: *member)(Args...);
-
         public:
+            MemberMethod(std::shared_ptr<Class> object, ReturnType(Class:: *method)(Args...)) noexcept
+                    : object(std::move(object)), method(method) { }
 
-            MemberMethod(std::shared_ptr<Class> object, ReturnType(Class:: *member)(Args...)) : object(object), member(member) { }
-
-            virtual ReturnType operator()(Args... args) override
+            ReturnType operator()(Args... args) override
             {
-                return (this->object.get()->*this->member)(std::forward<Args>(args)...);
+                return (*object.*method)(args...);
```

Comment:
Wow! Is it just simple as that? It works?

Reply:
yes, in c ++ it is not necessary to specify this-> for class fields. Given this fact, the code looks the same as before.

Reply:
Ok.